rfcs/records.md

```diff
@@ -0,0 +1,448 @@
+
+### Record type
+
+A record is a collectable object that stores the field values as a inline array of values (TValues) as well as a pointer to the shape. Shape is a table
+that stores various lookup data as implementation details as well as metafields.
+
+`type(r)` is `"record"`; `getmetatable(r)` can be used to retrieve the shape. `typeof(r)` is `$` followed by the name of the record as spelled in the source file.
+The prefix is required to make sure that builtin object types like `number` or host-provided userdata like `Vector3` can't be spoofed and confused with records.
+
+> TODO: Why `$`? :)
+
+Being a collectable object, records use raw equality by default when comparing using `==` or hashing; equality behavior can be overridden via `__eq`. From this
+perspective, one could think of records as a user-defined userdata type as opposed to a host-defined userdata type: both typically expose a strict set of fields,
+both are heap-allocated, both use contiguous storage.
+
+Shape contains field lookup data in extra storage that's only allocated for shape tables, as well as methods and metamethods stored as regular table entries.
+The field lookup data is internal and immutable; for example, it might contain a string->index dictionary to be able to quickly locate fields in internal storage.
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+The field lookup does not use `__index` or `__newindex` metamethods (or the metatable in general).
+
+Invoking methods with `:` desugars into `getmetatable(obj).method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the record has a given method as a field.
+
+> TODO: Should we use `__namecall` instead of raw MT access? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword.
+
+> TODO: The draft RFC suggests two options for the syntax; only one will be chosen in the final version
```

Comment:
Something to think about here is the ability to default a member to some value if one wasn't provided. This is something that you're capable of doing with the current OO design. Is this something we want to support as well?

```lua
local Stack = {}
Stack.__index = Stack

function Stack.new()
  local stack = {}
  stack.items = {} -- Empty array by default!
  return setmetatable(stack, Stack)
end

-- ... methods for Stack ...
```

I can imagine a syntax like

```lua
record Stack<T> = { items: {T} = {} }
-- ... methods for Stack ...
```

or 

```lua
record Stack<T>(items: {T} = {})
-- ... methods for Stack ...
```

... but the second one may require default arguments + named arguments to be usable.

Reply:
This is covered in the RFC I believe?

Reply:
Oh it is. Just further down. My bad!

Reply:
...no it's not. That's for `local person = Person { name = "hello" }`. I was talking about records themselves setting a default argument if the caller didn't provide it to begin with. I don't think I see that.

Reply:
> In variant A, it would make sense to allow omission of any field, in which case it gets replaced with the default of nil. A future extension (not part of this RFC) could be made to allow specification of default values at record definition time. Type checker would fail to type check record construction if fields that have non-optional types have the values omitted.

Reply:
(and further down there's the story for variant B for which I think it'd need to be a prefix of the arguments)

Reply:
FWIW I think I'm mostly set on variant A being the right variant. The only oddity there is that when the types are omitted, it results in a syntax that you don't exactly see today -- `{ x, y }` - but it seems easy to read and it's easy to parse so I don't think it's a big deal.

Reply:
Yeah, I fully agree that the braces syntax is the right way forward. It makes it more helpfully obvious that records are this thing that you can index on with the same name as its declaration. The latter looks extra weird because it looks like a function call syntax and the only way to get values out of it is to pattern match. Except that's not quite right because you also have argument names as properties.

Reply:
If i do

```lua
record Stack<T> = { items: {T} }
Stack.items = {}
```

Will `items` field get overwritten when `items` argument is omitted? Can you even do something like that?